lib/datadog/appsec/remote.rb

```diff
@@ -0,0 +1,126 @@
+
+        ASM_PRODUCTS = [
+          'ASM_DD',       # Datadog employee issued configuration
+          'ASM',          # customer issued configuration (rulesets, passlist...)
+          'ASM_FEATURES', # capabilities
+          'ASM_DATA',     # config files (IP addresses or users for blocking)
+        ].freeze
+
+        def capabilities
+          remote_features_enabled? ? ASM_CAPABILITIES : []
+        end
+
+        def products
+          remote_features_enabled? ? ASM_PRODUCTS : []
+        end
+
+        # rubocop:disable Metrics/MethodLength
+        def receivers
+          return [] unless remote_features_enabled?
+
+          matcher = Core::Remote::Dispatcher::Matcher::Product.new(ASM_PRODUCTS)
+          receiver = Core::Remote::Dispatcher::Receiver.new(matcher) do |repository, changes|
+            changes.each do |change|
+              Datadog.logger.debug { "remote config change: '#{change.path}'" }
+            end
+
+            rules = []
+            data = []
+            overrides = []
+            exclusions = []
+
+            asm_data_config_types = ['blocked_ips', 'blocked_users']
+            asm_overrides_config_types = ['blocking', 'disabled_rules']
+
+            repository.contents.each do |content|
+              case content.path.product
+              when 'ASM_DD'
+                rules << parse_content(content)
+              when 'ASM_DATA'
+                data << parse_content(content) if asm_data_config_types.include?(content.path.config_id)
+              when 'ASM'
+                overrides << parse_content(content) if asm_overrides_config_types.include?(content.path.config_id)
+                exclusions << parse_content(content) if content.path.config_id == 'exclusion_filters'
+              end
+            end
+
+            if rules.empty?
+              settings_rules = AppSec::Processor::RuleLoader.load_rules(ruleset: Datadog.configuration.appsec.ruleset)
+
+              raise NoRulesError, 'no default rules available' unless settings_rules
+
+              rules = [settings_rules]
+            end
+
+            ruleset = AppSec::Processor::RuleMerger.merge(
+              rules: rules,
+              data: data,
+              overrides: overrides,
+              exclusions: exclusions,
+            )
+
+            Datadog::AppSec.reconfigure(ruleset: ruleset)
+          end
+
+          [receiver]
+        end
+        # rubocop:enable Metrics/MethodLength
+
+        private
+
+        def remote_features_enabled?
+          # TODO: Only checking ENV is not enough, since customers could configure the ruleset via Datadog.configure
+          appsec_rules = ENV['DD_APPSEC_RULES']
+
+          return true unless appsec_rules
+
+          appsec_rules.empty?
+        end
+
+        def parse_content(content)
+          data = content.data.read
+
+          content.data.rewind
+
+          raise ReadError, 'EOF reached' if data.nil?
+
+          JSON.parse(data)
+        end
```

Comment:
Not to be fixed now, but I feel like this should be in `Core`, which hints me at something being amiss in terms of abstraction.

Reply:
Agree 100%. This method was the only one that didn't feel right when extracting everything, but I couldn't figure out where to place it. 

But as you mention, let's focus getting released and then refactor the code 😄